### PR TITLE
Add CP_MDA-exact algorithm for correct uncertainty estimates with missing data.

### PR DESCRIPTION
**Pull request by Fdvanleeuwen**
The CP_MDA-exact algorithm [1] allows users to obtain conformalised uncertainty estimates for TabPFNRegressor. This allows users to have correct uncertainty estimates based on the missing masks, as described in [2]. There are two functions:

CP_MDA_TabPFNRegressor: for fitting TabPFNRegressor and obtaining the correction terms for each mask.
CP_MDA_TabPFNRegressor_newdata: for applying the fitted model (and correction terms) to new test cases.

[1] Margaux Zaffran, Aymeric Dieuleveut, Julie Josse, and Yaniv Romano. Conformal prediction with missing
values. In International Conference on Machine Learning, pages 40578–40604. PMLR, 2023.
[2] van Leeuwen, F. D. Conformal Prediction for Tabular Prior-Data Fitted Networks with Missing data. In EurIPS 2025 Workshop: AI for Tabular Data.

### CODE REVIEW

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            
+            # add the mask itself
+            nested_masks_with_self = nested_masks + [i]  # Create new list instead of append
+
+            # obtain indexes for the rows
+            indexes = df_with_ids[df_with_ids["mask_id"].isin(nested_masks_with_self)].index
+
+            # select the validation data based on the indices
+            X_val_nested = self.X_val.iloc[indexes]
+            Y_val_nested = self.Y_val.iloc[indexes]
```
[the remark this answers is not in the document]
Comment (Fdvanleeuwen) [reply src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: Not sure if this is needed if merge preserves the order of the rows in Mask_val, but I will change it anyway.

File: examples/CP_missing_data/CP_missing_data_example.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import numpy as np
+import pandas as pd
+import warnings
+
+try:
+    # Try standard TabPFN package first
+    from tabpfn import TabPFNRegressor
+except ImportError:
+    # Fall back to TabPFN client
+    from tabpfn_client import  TabPFNRegressor
```
Comment (bejaeger) [inline examples/CP_missing_data/CP_missing_data_example.py]: You can use `from tabpfn_extensions.utils import TabPFNClassifier, TabPFNRegressor` instead :)

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+import numpy as np
+import pandas as pd
+from sklearn.model_selection import train_test_split
+
+try:
+    # Try standard TabPFN package first
+    from tabpfn import TabPFNRegressor
+except ImportError:
+    # Fall back to TabPFN client
+    from tabpfn_client import  TabPFNRegressor
```
Comment (bejaeger) [inline src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: Same here: `from tabpfn_extensions.utils import TabPFNClassifier, TabPFNRegressor`

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+
+class CP_MDA_TabPFNRegressor:
+    """
+    Compute the correction terms for missing data masks using conformal prediction.
+
+    Parameters:
+        X_train : matrix-like of shape (n_samples, n_predictors)
+
+        Y_train : array-like of continuous outcome with shape (n_samples,)
```
Comment (bejaeger) [inline src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: Let's stick to the convention of parsing the data during the `fit()` call. Only configuration should go here.

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+                "Some masks have very small calibration sets")
+
+        correction_term = np.quantile(s, Q_use)
+        return correction_term
+
+    def split_data(self):
+        """Split data into training and validation sets."""
+        # create df with missing data indicator
+        missing_bool_df = self.X.isnull().astype(int)
+        self.X_train, self.X_val, Y_train_arr, Y_val_arr, self.Mask_train, self.Mask_val = train_test_split(
```
Comment (bejaeger) [inline src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: Our convention is to use `y` for the target variable (lowercase).
Please also make `Mask_train/val` lowercase -> `mask_train/val` to stick to python conventions

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+
+        return self.mask_unique, self.model
+
+    def fit(self):
+        """Convenience method to run the entire pipeline"""
+        self.split_data()
+        self.run_TABPFN()
+        self.mask_preprocess()
+        mask_unique, model = self.create_calibration_sets()
```
Comment (bejaeger) [inline src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: It would be easier to follow the code if you return the variables needed in the next steps rather than making them an instance variable. Would be a nice change.

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+class CP_MDA_TabPFNRegressor_newdata:
+    """
+    Compute the correction terms for missing data masks using conformal prediction.
+
+    Parameters:
+
+    TabPFN: Fitted TabPFNRegressor model.
+
+    X_new : matrix-like of shape (n_samples, n_predictors)
```
Comment (bejaeger) [inline src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: Same here, let's parse `X_new` that during `fit()`

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        lb_corr = preds_test[0] - self.mask_test_cor["correction_term"].values
+        ub_corr = preds_test[2] + self.mask_test_cor["correction_term"].values
+
+        return lb_corr, preds_test[1], ub_corr, preds_test[0], preds_test[2]
+
+    def fit(self):
+        """Convenience method to run the entire pipeline"""
+        self.obtain_preds()
+        self.match_mask()
+        CP_results =  self.perform_correction()
```
Comment (bejaeger) [inline src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: similar here, could we return and parse the arguments needed?

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+
+
+    Returns:
+     mask_unique: DataFrame with the correction terms for each mask.
+
+     model: Fitted TabPFNRegressor model.
+
+    """
+
+    def __init__(self, X_train, Y_train, quantiles, val_size, seed=None):
```
Comment (bejaeger) [inline src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: Please add type hints here and in all other functions.

**Review by bejaeger — CHANGES_REQUESTED** [review]
Hi @Fdvanleeuwen ,
thank you so much for this contribution. It looks great all in all. There are a few stylistic things I commented on that would be great if you could change them. The CI also shows some errors. If you could fix these I can have a closer look at the method.
Thanks again!

**Comment by noahho** [comment]
Hi @Fdvanleeuwen — this PR has been quiet for 101 days but reviewer feedback is above. Could you take a look when you get a chance? Happy to land it or close it out, whichever fits your plans.

**Comment by Fdvanleeuwen** [comment]
> Hi @Fdvanleeuwen — this PR has been quiet for 101 days but reviewer feedback is above. Could you take a look when you get a chance? Happy to land it or close it out, whichever fits your plans.

Sorry, forgot to request a review of the changes. Hope it works now.

**Comment by noahho** [comment]
@bejaeger — friendly nudge, @Fdvanleeuwen pushed changes and re-requested review on May 18 addressing your earlier feedback. Could you take another look when you have a moment? Thanks!

File: examples/cp_missing_data/cp_missing_data_example.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+print(unique_patterns)
+
+# Use TabPFN+CP-MDA, the interval 
+# Note the interval needs to be symmetric
+model = CPMDATabPFNRegressor(quantiles=[0.05, 0.5, 0.95], val_size=0.5, seed=123)
+calibration_results, model_fit = model.fit(X, y)
+print(calibration_results)
+
+# Apply the model to new cases 
+cp_apply = CPMDATabPFNRegressorNewData(model_fit, quantiles=[0.05, 0.5, 0.95], calibration_results=calibration_results)
```
Comment (bejaeger) [inline examples/cp_missing_data/cp_missing_data_example.py]: Instead of the 2-class approach, could you use a standard sklearn API similar to the other methods in this repo?
You can collapse into one `CPMDATabPFNRegressor(BaseEstimator, RegressorMixin)` that stores `calibration_results_` and `model_` as fitted attributes and exposes .predict(X).
That removes the two-class approach and makes it composable with `Pipeline`, `cloning`, `get_params`, etc.

File: src/tabpfn_extensions/cp_missing_data/cp_missing_data.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+            x, y, missing_bool_df, test_size=self.val_size, random_state=self.seed
+        )
+
+        # Convert y arrays back to pandas Series to maintain .iloc functionality
+        y_train = pd.Series(y_train_arr, index=x_train.index)
+        y_val = pd.Series(y_val_arr, index=x_val.index)
+
+        return x_train, x_val, y_train, y_val, mask_train, mask_val
+
+    def run_TABPFN(self, x_train: pd.DataFrame, y_train: pd.Series) -> TabPFNRegressor:
```
Comment (bejaeger) [inline src/tabpfn_extensions/cp_missing_data/cp_missing_data.py]:
```suggestion
    def run_tabpfn(self, x_train: pd.DataFrame, y_train: pd.Series) -> TabPFNRegressor:
```

File: src/tabpfn_extensions/cp_missing_data/cp_missing_data.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+    Returns:
+        mask_unique: DataFrame with the correction terms for each mask.
+        model: Fitted TabPFNRegressor model.
+    """
+
+    def __init__(
+        self, quantiles: list[float], val_size: float, seed: int | None = None
+    ) -> None:
+        self.quantiles = quantiles
```
Comment (bejaeger) [inline src/tabpfn_extensions/cp_missing_data/cp_missing_data.py:62]: can you assert the shape of `quantiles` is correct?

File: src/tabpfn_extensions/cp_missing_data/cp_missing_data.py [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+
+    def predict(
+        self, x_new: ArrayLike
+    ) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        """Convenience method to run the entire pipeline.
+
+        Parameters:
+            x_new : matrix-like of shape (n_samples, n_predictors)
+        """
+        x = pd.DataFrame(x_new)
```
Comment (bejaeger) [inline src/tabpfn_extensions/cp_missing_data/cp_missing_data.py:308]: this needs to have the same format as in fit. Can you assert the column names match the ones passed in fit if a DataFrame is provided?

File: src/tabpfn_extensions/cp_missing_data/cp_missing_data.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+        self.quantiles = quantiles
+        self.val_size = val_size
+        self.alpha = quantiles[0] * 2
+        self.seed = seed
+
+    def calc_correction_term(
+        self,
+        predictions: tuple[np.ndarray, np.ndarray, np.ndarray],
+        y_val: pd.Series,
+        alpha: float,
```
Comment (bejaeger) [inline src/tabpfn_extensions/cp_missing_data/cp_missing_data.py]: only called with `self.alpha` so the `alpha` argument is not needed.

File: src/tabpfn_extensions/cp_missing_data/cp_missing_data.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+        # check if there are masks in the test set that are not in the calibration set
+        new_masks = mask_test_cor[mask_test_cor["correction_term"].isna()][mask_cols]
+
+        if new_masks.shape[0] > 0:
+            warnings.warn(
+                "The following masks are not in the calibration set:\n"
+                f"{new_masks.to_string()}\n"
+                "The baseline quantile estimates will be returned for those cases.",
+                stacklevel=2,
+            )
```
Comment (bejaeger) [inline src/tabpfn_extensions/cp_missing_data/cp_missing_data.py:283]: This warning says that the baseline quantile estimates are returned, but that does not seem to be true.

Below in `lb_corr = preds[0] - mask_test_cor["correction_term"].values` the `lb_corr` would become NaN IIUC. Could you fix this please?

**Review by bejaeger — CHANGES_REQUESTED** [review]
Hi @Fdvanleeuwen ,
thanks again for working on this and sorry for the slow reply.

The PR looks almost ready to go, there are a few things that would need some attention from your side. Thanks!

File: src/tabpfn_extensions/cp_missing_data/cp_missing_data.py (hunk 15 above)
Comment (Fdvanleeuwen) [reply src/tabpfn_extensions/cp_missing_data/cp_missing_data.py:283]: very nice catch

File: src/tabpfn_extensions/CP_missing_data/CP_missing_data.py [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+        )
+
+        # Convert Y arrays back to pandas Series to maintain .iloc functionality
+        self.Y_train = pd.Series(Y_train_arr, index=self.X_train.index)
+        self.Y_val = pd.Series(Y_val_arr, index=self.X_val.index)
+
+    def run_TABPFN(self):
+        """Fit the TabPFN model."""
+        # fit model
+        m_fit = TabPFNRegressor()
```
[the remark this answers is not in the document]
Comment (bejaeger) [reply src/tabpfn_extensions/CP_missing_data/CP_missing_data.py]: This could be a good addition in a follow-up PR.

**Review by bejaeger — APPROVED** [review]
Hi @Fdvanleeuwen , thanks a lot. There are a few comments unaddressed but I'll approve for now. Would be great if you could have a look.

Thanks again for this contribution!

**Comment by adrian-prior** [comment]
@Fdvanleeuwen, I would love to merge this PR soon! Did you already have a look at @bejaeger's comments? Also, it looks like a changelog is missing (I recently added this as a requirement). Could you add that?

**Comment by Fdvanleeuwen** [comment]
> @Fdvanleeuwen, I would love to merge this PR soon! Did you already have a look at @bejaeger's comments? Also, it looks like a changelog is missing (I recently added this as a requirement). Could you add that?

@adrian-prior The tabpfn_estimator parameter has been added, allowing users to pass a pre-configured TabPFNRegressor with custom settings such as n_estimators or device. This was the only comment left open, I believe. Also, the change log is added :)
